TestNewBatchSpanProcessorWithOptions/non-default_ExportTimeout in go.opentelemetry.io/otel/sdk/trace fails sporadically

### Description

https://github.com/open-telemetry/opentelemetry-go/pull/1978/checks?check_run_id=2778145087

```
--- FAIL: TestNewBatchSpanProcessorWithOptions (0.04s)
    --- FAIL: TestNewBatchSpanProcessorWithOptions/non-default_ExportTimeout (0.02s)
        batch_span_processor_test.go:236: context deadline: got err <nil>, want context deadline exceeded
FAIL
FAIL	go.opentelemetry.io/otel/sdk/trace	0.314s
ok  	go.opentelemetry.io/otel/sdk/trace/tracetest	0.117s
FAIL
make: *** [test] Error 1
Error: Process completed with exit code 2.
```

It looks like the delay that is setup in the mock exporter is not sufficient to trigger the expected error.

https://github.com/open-telemetry/opentelemetry-go/blob/cbcd4b1a3dc6ff839e06749c83294260aa671e6f/sdk/trace/batch_span_processor_test.go#L139

A different approach should be explored to test this and the dueling timers approach should be removed.

### Environment

- OS: Mac OS X 10.15.7
- Architecture: amd64
- Go Version: 1.16
- opentelemetry-go version: d7778a6 

### Steps To Reproduce

1. Use the configuration '...'
2. Run '...'
3. See error

### Test Failure Logs

[log.txt](https://github.com/open-telemetry/opentelemetry-go/files/6619762/log.txt)

--- a/sdk/trace/batch_span_processor_test.go
+++ b/sdk/trace/batch_span_processor_test.go
@@ -36,7 +36,6 @@ type testBatchExporter struct {
 	sizes         []int
 	batchCount    int
 	shutdownCount int
-	delay         time.Duration
 	errors        []error
 	droppedCount  int
 	idx           int
@@ -53,8 +52,6 @@ func (t *testBatchExporter) ExportSpans(ctx context.Context, spans []sdktrace.Re
 		t.idx++
 		return err
 	}
-
-	time.Sleep(t.delay)
 
 	select {
 	case <-ctx.Done():
@@ -109,34 +106,22 @@ func TestNewBatchSpanProcessorWithNilExporter(t *testing.T) {
 }
 
 type testOption struct {
-	name              string
-	o                 []sdktrace.BatchSpanProcessorOption
-	wantNumSpans      int
-	wantBatchCount    int
-	wantExportTimeout bool
-	genNumSpans       int
-	delayExportBy     time.Duration
-	parallel          bool
+	name           string
+	o              []sdktrace.BatchSpanProcessorOption
+	wantNumSpans   int
+	wantBatchCount int
+	genNumSpans    int
+	parallel       bool
 }
 
 func TestNewBatchSpanProcessorWithOptions(t *testing.T) {
 	schDelay := 200 * time.Millisecond
-	exportTimeout := time.Millisecond
 	options := []testOption{
 		{
 			name:           "default BatchSpanProcessorOptions",
 			wantNumSpans:   2053,
 			wantBatchCount: 4,
 			genNumSpans:    2053,
-		},
-		{
-			name: "non-default ExportTimeout",
-			o: []sdktrace.BatchSpanProcessorOption{
-				sdktrace.WithExportTimeout(exportTimeout),
-			},
-			wantExportTimeout: true,
-			genNumSpans:       2053,
-			delayExportBy:     2 * exportTimeout, // to ensure export timeout
 		},
 		{
 			name: "non-default BatchTimeout",
@@ -204,9 +189,7 @@ func TestNewBatchSpanProcessorWithOptions(t *testing.T) {
 	}
 	for _, option := range options {
 		t.Run(option.name, func(t *testing.T) {
-			te := testBatchExporter{
-				delay: option.delayExportBy,
-			}
+			te := testBatchExporter{}
 			tp := basicTracerProvider(t)
 			ssp := createAndRegisterBatchSP(option, &te)
 			if ssp == nil {
@@ -231,12 +214,38 @@ func TestNewBatchSpanProcessorWithOptions(t *testing.T) {
 					gotBatchCount, option.wantBatchCount)
 				t.Errorf("Batches %v\n", te.sizes)
 			}
-
-			if option.wantExportTimeout && te.err != context.DeadlineExceeded {
-				t.Errorf("context deadline: got err %+v, want %+v\n",
-					te.err, context.DeadlineExceeded)
-			}
 		})
+	}
+}
+
+type stuckExporter struct {
+	testBatchExporter
+}
+
+// ExportSpans waits for ctx to expire and returns that error.
+func (e *stuckExporter) ExportSpans(ctx context.Context, _ []sdktrace.ReadOnlySpan) error {
+	<-ctx.Done()
+	e.err = ctx.Err()
+	return ctx.Err()
+}
+
+func TestBatchSpanProcessorExportTimeout(t *testing.T) {
+	exp := new(stuckExporter)
+	bsp := sdktrace.NewBatchSpanProcessor(
+		exp,
+		// Set a non-zero export timeout so a deadline is set.
+		sdktrace.WithExportTimeout(1*time.Microsecond),
+		sdktrace.WithBlocking(),
+	)
+	tp := basicTracerProvider(t)
+	tp.RegisterSpanProcessor(bsp)
+
+	tr := tp.Tracer("BatchSpanProcessorExportTimeout")
+	generateSpan(t, false, tr, testOption{genNumSpans: 1})
+	tp.UnregisterSpanProcessor(bsp)
+
+	if exp.err != context.DeadlineExceeded {
+		t.Errorf("context deadline error not returned: got %+v", exp.err)
 	}
 }
 

